x-pack/plugins/actions/server/actions_client.ts

```diff
@@ -0,0 +1,174 @@
+    perPage?: number;
+    page?: number;
+    search?: string;
+    defaultSearchOperator?: 'AND' | 'OR';
+    searchFields?: string[];
+    sortField?: string;
+    hasReference?: {
+      type: string;
+      id: string;
+    };
+    fields?: string[];
+  };
+}
+
+interface ConstructorOptions {
+  actionTypeService: ActionTypeService;
+  encryptedSavedObjectsPlugin: EncryptedSavedObjectsPlugin;
+  savedObjectsClient: SavedObjectsClient;
+}
+
+interface UpdateOptions {
+  id: string;
+  data: Action;
+  options: { version?: string; references?: SavedObjectReference[] };
+}
+
+export class ActionsClient {
+  private savedObjectsClient: SavedObjectsClient;
+  private actionTypeService: ActionTypeService;
+  private encryptedSavedObjectsPlugin: EncryptedSavedObjectsPlugin;
+
+  constructor({
+    actionTypeService,
+    encryptedSavedObjectsPlugin,
+    savedObjectsClient,
+  }: ConstructorOptions) {
+    this.actionTypeService = actionTypeService;
+    this.encryptedSavedObjectsPlugin = encryptedSavedObjectsPlugin;
+    this.savedObjectsClient = savedObjectsClient;
+  }
+
+  /**
+   * Create an action
+   */
+  public async create({ data, options }: CreateOptions) {
+    const { actionTypeId } = data;
+    if (!this.actionTypeService.has(actionTypeId)) {
+      throw Boom.badRequest(
+        i18n.translate('xpack.actions.actionsClient.create.missingActionTypeError', {
+          defaultMessage: 'Action type "{actionTypeId}" is not registered.',
+          values: {
+            actionTypeId,
+          },
+        })
+      );
+    }
+    this.actionTypeService.validateActionTypeConfig(actionTypeId, data.actionTypeConfig);
+    const actionWithSplitActionTypeConfig = this.moveEncryptedAttributesToSecrets(data);
+    return await this.savedObjectsClient.create('action', actionWithSplitActionTypeConfig, options);
+  }
+
+  /**
+   * Get an action
+   */
+  public async get({ id }: { id: string }) {
+    return await this.savedObjectsClient.get('action', id);
+  }
+
+  /**
+   * Find actions
+   */
+  public async find({ options = {} }: FindOptions) {
+    return await this.savedObjectsClient.find({
+      ...options,
+      type: 'action',
+    });
+  }
+
+  /**
+   * Delete action
+   */
+  public async delete({ id }: { id: string }) {
+    return await this.savedObjectsClient.delete('action', id);
+  }
+
+  /**
+   * Update action
+   */
+  public async update({ id, data, options = {} }: UpdateOptions) {
+    const { actionTypeId } = data;
+    if (!this.actionTypeService.has(actionTypeId)) {
+      throw Boom.badRequest(`Action type "${actionTypeId}" is not registered.`);
+    }
+    if (data.actionTypeConfig) {
+      this.actionTypeService.validateActionTypeConfig(actionTypeId, data.actionTypeConfig);
+      data = this.moveEncryptedAttributesToSecrets(data);
+    }
+    return await this.savedObjectsClient.update<any>('action', id, data, options);
+  }
+
+  /**
+   * Fire an action
+   */
+  public async fire({ id, params }: FireOptions) {
```

Comment:
This came across my mind at one point. I think I skipped the endpoint and was planning to develop it as needed. The main use case that came to my mind is being called from alerting / task manager services. Adding a basic endpoint is definitely something that can be done.

Reply:
Should we consider moving `fire` out of the `ActionsClient` to a different class? All of the other methods on the `ActionsClient` are safe to consume within route handlers while ignoring security (because the saved objects client enforces it), but the `fire` method is using the internal server user to decrypt the attributes and actually fire the action, hence `getDecryptedAsInternalUser`. If we do wish for `fire` to be able to triggered by the end user from an api request, we'll need to ensure that we aren't returning sensitive information. Having "safe" and "potentially unsafe" operations on the same class could lead to potential problems long-term.